cirq-core/cirq/ops/state_preparation_channel.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Quantum gates to prepare a given target state."""
+
+from typing import Any, Dict, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import protocols
+from cirq.ops import raw_types
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class StatePreparationChannel(raw_types.Gate):
+    """A channel which prepares any state provided as the state vector on it's target qubits."""
+
+    def __init__(self, target_state: np.ndarray, name: str = "StatePreparation") -> None:
+        """Initializes a State Preparation channel.
+
+        Args:
+            target_state: The state vector that this gate should prepare.
+            name: the name of the gate
+
+        Raises:
+            ValueError: if the array is not 1D, or does not have 2**n elements for some integer n.
+        """
+        if len(target_state.shape) != 1:
+            raise ValueError('`target_state` must be a 1d numpy array.')
+
+        n = int(np.round(np.log2(target_state.shape[0] or 1)))
+        if 2 ** n != target_state.shape[0]:
+            raise ValueError(f'Matrix width ({target_state.shape[0]}) is not a power of 2')
+
+        self._state = target_state.astype(np.complex128) / np.linalg.norm(target_state)
+        self._num_qubits = n
+        self._name = name
+        self._qid_shape = (2,) * n
+
+    @staticmethod
```

Comment:
We usually always define `_has_unitary_` and `_has_kraus_` as instance methods and not static methods. I'd prefer to keep things consistent across the codebase, unless there's a strong reason not to?

Reply:
PyCharm keeps suggesting to make it a staticmethod, is the some config used by Cirq developers that fits the cirq style? Anyways, made it non-static for now.